Try: Add synced patterns to theme on save

mikachan opened the pull request:
This PR explores the most minimal way we can introduce some pattern management into the plugin by allowing custom patterns to be saved to the theme filesystem on save, allowing them to be exported with the theme.

Part of https://github.com/WordPress/create-block-theme/issues/287.

<img width="288" alt="image" src="https://github.com/WordPress/create-block-theme/assets/1645628/557380f9-d141-4542-bb18-915d836d4aa7">

## Testing Instructions
1. Create a new synced pattern in the Editor (also try adding categories)
2. Go to the Editor, and in the Create Block Theme menu, save the theme changes with the "Save My Patterns" option checked

mikachan committed:
Add savePatterns option

mikachan committed:
Add basic add_patterns_to_theme function

mikachan committed:
Switch order of save options

mikachan committed:
Add prepare_pattern_for_export function

mikachan committed:
Copy media to theme filesystem

mikachan committed:
Fix pattern slug

mikachan committed:
Add replace_local_pattern_references

mikachan committed:
Tidy up pattern_from_wp_block

mikachan committed:
Add pattern categories

mikachan committed:
Default to empty string for categories list

mikachan committed:
Potentially save pattern sync status

eirichmond commented:
@mikachan Great job on this! Initial testing works well. I know it's a work in progress, but I wondered why the Inserter was set to 'no' in the pattern header. Once saved and exported, the pattern still doesn't come from the theme/patterns folder; it remains in the database. I guess there is logic behind that. For me, if I'm developing a theme, I wouldn't want anything left in the database after saving and exporting the patterns. I think there might be some thinking behind synced patterns, but as a developer creating patterns as part of the theme, I'd want them not to be restricted and synced. I'd prefer the theme to be distributed to users who could then decide to generate their own synced patterns from the ones provided by the theme, if that makes sense.

Also, I think the heredoc approach might be more readable for generating the header. This allows you to include the variables directly within the string without needing to concatenate multiple parts.

Again, I know this is a work in progress, so I guess any image-related details (e.g., downloaded to the theme assets/images and changing the reference to get_theme_file_uri('xxxx'), etc.) haven't been addressed yet. I'd love to contribute to this with my own pull request if it helps. Great job 🥳

eirichmond commented:
In hindsight, I guess the reasoning for not removing them from the database is to preserve synced patterns that have already been placed throughout the site. It would be really useful if there was a flag saved within postmeta indicating whether a pattern was already synced. This could then be used to avoid removing the pattern when the theme is saved and the patterns are exported.

mikachan commented:
> It would be really useful if there was a flag saved within postmeta indicating whether a pattern was already synced.

It looks like the sync status is saved in postmeta under `wp_pattern_sync_status`, and you're able to access it using:

`get_post_meta( $pattern->id, 'wp_pattern_sync_status', true );`

mikachan committed:
Merge branch 'trunk' into try/export-patterns

mikachan commented on includes/create-theme/theme-patterns.php:
```diff
@@ -21,6 +21,30 @@ public static function pattern_from_template( $template, $new_slug = null ) {
+		$pattern->slug         = wp_get_theme()->get( 'TextDomain' ) . '/' . $pattern->name;
+		$pattern_category_list = get_the_terms( $pattern->id, 'wp_pattern_category' );
+		$pattern->categories   = ! empty( $pattern_category_list ) ? join( ', ', wp_list_pluck( $pattern_category_list, 'name' ) ) : '';
+		// $pattern->sync_status  = get_post_meta( $pattern->id, 'wp_pattern_sync_status', true );
```
I'm not sure this belongs here but I've left it here commented out as a reference for now.

mikachan committed:
Merge branch 'trunk' into try/export-patterns

mikachan committed:
Merge branch 'trunk' into try/export-patterns

# Conflicts:
#	includes/class-create-block-theme-api.php

mikachan committed:
Refactor PHP content of pattern_from_template

Co-Authored-By: Elliott Richmond <elliott@squareonemd.co.uk>

mikachan committed:
Refactor PHP content of pattern_from_wp_block

Co-Authored-By: Elliott Richmond <elliott@squareonemd.co.uk>

mikachan committed:
Delete synced patterns after adding to theme

Co-Authored-By: Elliott Richmond <elliott@squareonemd.co.uk>

mikachan committed:
Refactor and add error handling

mikachan commented:
I've updated this PR with the changes from @eirichmond's PR (https://github.com/WordPress/create-block-theme/pull/693), refactored a couple of things and added some error handling. I think this is ready for review 🤞

jffng commented on includes/create-theme/theme-patterns.php:
```diff
@@ -47,4 +71,111 @@ public static function create_pattern_link( $attributes ) {
+					// Create the pattern file.
+					$pattern_file = $patterns_dir . 'pattern-' . $pattern->name . '.php';
+					file_put_contents(
+						$patterns_dir . DIRECTORY_SEPARATOR . 'pattern-' . $pattern->name . '.php',
```
Do you think the `pattern-` prefix is necessary?

```suggestion
						$patterns_dir . DIRECTORY_SEPARATOR . 'pattern-' . $pattern->name . '.php',
```

colorful-tones replied:
This threw me off. I could not find my patterns and thought it was not working ...until I saw this hint:

> Do you think the pattern- prefix is necessary?

jffng replied:
Yeah same! Especially if the theme already has lots of patterns it was hard to spot.

jffng replied:
I think I expected the file name to be the kebab-cased pattern name.

mikachan replied:
I'm not sure why I added the prefix, haha 🙈 I'll remove it for now, I can't see what value it's adding. If I remember I'll let you know!

mikachan replied:
Removed the prefix here: https://github.com/WordPress/create-block-theme/pull/675/commits/f017b9485cfddac0bc0f0f5cf2f5312304f42c4e.

jffng commented on includes/create-theme/theme-patterns.php:
```diff
@@ -47,4 +71,111 @@ public static function create_pattern_link( $attributes ) {
+					self::replace_local_pattern_references( $pattern );
+
+					// Remove it from the database to ensure that these patterns are loaded from the theme.
+					wp_delete_post( $pattern->id, true );
```
If I create a new pattern and save it using CBT, it reloads the editor but the post has been deleted. With Gutenberg active I at least get a better error response: 

**Without Gutenberg**
https://github.com/user-attachments/assets/79a8cd6a-6f0a-4a5e-9339-228b3b8da0e3

**Gutenberg 19.0.0**
 ![Screenshot 2024-08-14 at 2 49 49 PM](https://github.com/user-attachments/assets/37537f1d-41f2-431e-882b-9b52493bb0cb)

I wonder if we can handle this redirection better, maybe return the user back to the patterns view?

colorful-tones replied:
I received same error with latest Gutenberg trunk. When I disabled Gutenberg I got WSOD with HTML `<title>null ‹ Template ‹ Gutenberg Trunk ‹ Editor — WordPress</title>` 🤔

mikachan replied:
> I wonder if we can handle this redirection better, maybe return the user back to the patterns view?

Yes, nice idea, I've done this here: https://github.com/WordPress/create-block-theme/pull/675/commits/6de813f38bcce5951cf35ff6ac33d36cda4a6318.

Thanks both!

jffng reviewed (commented):
This is testing pretty well for me, nice work! I was able to save custom patterns and it exported with the theme.

mikachan committed:
Update save patterns option description

mikachan committed:
Remove pattern- prefix

mikachan committed:
Update option description

mikachan committed:
Redirect to patterns page if editing a pattern

mikachan committed:
Add check for preference.savePatterns

mikachan commented:
Thanks for the feedback @jffng and @colorful-tones! This should be ready for another review.

colorful-tones commented:
I'm confused about something in my testing. I created a synced pattern, then assigned it to a template, and then went into CBT and checked _only_ the 'Save Synced Patterns' thinking that the template would also be saved, because of this overall statement: "Any synced patterns created in the Editor will be moved to the theme. Note that this will delete all synced patterns from the Editor and any references in templates will be made relative to the theme." 

However, after hitting the 'Save Changes' from CBT I only see my new pattern in the patterns directory and the template has no changes at all, which makes me think the statement is a bit misleading. I imagine I would want to check the "Save Template Changes" box too, but the "any references in templates will be made relative to the theme." next to the "Save Synced Patterns" implied to me that my template changes would be saved too. 🤔 

Also, after saving patterns to the theme when I go back to the template it states: "Block has been deleted or is unavailable."

![Screenshot 2024-08-15 at 10 44 34 AM](https://github.com/user-attachments/assets/e5283d01-509b-41aa-9a63-f0a9b0a2dca0)

![Screenshot 2024-08-15 at 10 49 30 AM](https://github.com/user-attachments/assets/f251463d-0088-4261-a7b9-2776300971ac)

mikachan committed:
Update templates that reference the pattern

mikachan committed:
Merge branch 'trunk' into try/export-patterns

mikachan commented:
Thanks @colorful-tones! I've pushed an update that will update any templates that reference the pattern and save those changes to the theme, regardless of whether the "Save Template Changes" option is checked. I think that matches up with the expectations in the "Save Sycned Patterns" description. This should also fix the "Block has been deleted or is unavailable." issue.

colorful-tones commented:
@mikachan I did a bunch of testing and everything is working. Nice work! 

I think this functionality will help folks with creating and saving their patterns.

colorful-tones reviewed (approved):
LGTM! I think this functionality will be super helpful.

mikachan commented:
Thanks so much for reviewing, @colorful-tones!
